metadata/src/main/java/org/apache/kafka/controller/FeatureControlManager.java

```diff
@@ -232,13 +248,20 @@ private ApiError updateMetadataVersion(
         Consumer<ApiMessageAndVersion> recordConsumer
     ) {
         MetadataVersion currentVersion = metadataVersion();
+        ZkMigrationState zkMigrationState = zkMigrationState();
         final MetadataVersion newVersion;
         try {
             newVersion = MetadataVersion.fromFeatureLevel(newVersionLevel);
         } catch (IllegalArgumentException e) {
             return invalidMetadataVersion(newVersionLevel, "Unknown metadata.version.");
         }
 
+        // Don't allow metadata.version changes while we're migrating
+        if (EnumSet.of(ZkMigrationState.PRE_MIGRATION, ZkMigrationState.MIGRATION).contains(zkMigrationState)) {
```

Comment:
Hmm, seems like this might come up in a bunch of places. Maybe we can have something like `boolean ZkMigrationState#inProgress() { return this == PRE_MIGRATION || this == POST_MIGRATION; }`?